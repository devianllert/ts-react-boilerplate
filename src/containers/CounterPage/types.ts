export interface CounterState {
  count: number;
}

export interface CounterIncrement {
  type: 'COUNTER_INCREMENT';
}

export interface CounterDecrement {
  type: 'COUNTER_DECREMENT';
}

export interface CounterClear {
  type: 'COUNTER_CLEAR';
}

export type CounterActions = CounterIncrement | CounterDecrement | CounterClear;
