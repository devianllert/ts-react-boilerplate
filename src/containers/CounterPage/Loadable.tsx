import loadable, { ImportComponent } from '../../utils/loadable';

const LoadableCounter = loadable((): ImportComponent => import(/* webpackChunkName: "counterPage" */ './CounterPage'));

export default LoadableCounter;
