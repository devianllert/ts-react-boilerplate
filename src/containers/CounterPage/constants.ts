export const COUNTER_INCREMENT = 'boilerplate/Counter/COUNTER_INCREMENT';
export const COUNTER_DECREMENT = 'boilerplate/Counter/COUNTER_DECREMENT';
export const COUNTER_CLEAR = 'boilerplate/Counter/COUNTER_CLEAR';
