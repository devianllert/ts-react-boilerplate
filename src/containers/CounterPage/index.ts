import Counter from './CounterPage';

export default Counter;
