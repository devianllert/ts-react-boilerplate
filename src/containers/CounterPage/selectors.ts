import { createSelector, OutputSelector } from 'reselect';

type CounterState = import('./types').CounterState;

const selectCounter = (state: AppState): CounterState => state.counter;

const makeSelectCount = (): OutputSelector<AppState, number, (res: CounterState) => number> => createSelector(
  selectCounter,
  (counter): number => counter.count,
);

export { selectCounter, makeSelectCount };
