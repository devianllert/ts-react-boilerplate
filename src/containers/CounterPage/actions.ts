import { CounterIncrement, CounterDecrement, CounterClear } from './types';
import { COUNTER_INCREMENT, COUNTER_DECREMENT, COUNTER_CLEAR } from './constants';

const counterIncrement = (): CounterIncrement => ({
  type: COUNTER_INCREMENT,
});

const counterDecrement = (): CounterDecrement => ({
  type: COUNTER_DECREMENT,
});

const counterClear = (): CounterClear => ({
  type: COUNTER_CLEAR,
});

export {
  counterIncrement,
  counterDecrement,
  counterClear,
};
