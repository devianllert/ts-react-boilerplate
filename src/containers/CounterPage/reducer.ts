import { Reducer } from 'redux';

import { COUNTER_INCREMENT, COUNTER_DECREMENT, COUNTER_CLEAR } from './constants';
import { CounterState, CounterActions } from './types';

const initialState: CounterState = {
  count: 0,
};

const counterReducer: Reducer<CounterState, CounterActions> = (state = initialState, action): CounterState => {
  switch (action.type) {
    case COUNTER_INCREMENT:
      return {
        ...state,
        count: state.count + 1,
      };
    case COUNTER_DECREMENT:
      return {
        ...state,
        count: state.count - 1,
      };
    case COUNTER_CLEAR:
      return {
        ...state,
        count: 0,
      };
    default:
      return state;
  }
};

export default counterReducer;
