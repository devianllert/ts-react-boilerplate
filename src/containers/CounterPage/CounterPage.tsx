import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { useTranslation } from 'react-i18next';

import useKey from '../../hooks/useKey';

import { makeSelectCount } from './selectors';

import { counterIncrement, counterDecrement } from './actions';

import Button from '../../components/Button';
import Tooltip from '../../components/Tooltip';

import styles from './CounterPage.module.scss';

interface StateToProps {
  count: number;
}

interface DispatchToProps {
  increment: () => import('./types').CounterIncrement;
  decrement: () => import('./types').CounterDecrement;
}

type Props = StateToProps & DispatchToProps & {};

const CounterPage = ({ count, increment, decrement }: Props): ReactElement => {
  const { t } = useTranslation();

  useKey('ArrowUp', increment);
  useKey('ArrowDown', decrement);

  return (
    <>
      <Helmet>
        <title>{t('counter.headTitle')}</title>
      </Helmet>

      <div className={styles.content}>
        <div className={styles.counter}>
          <Tooltip title="Decrement" placement="top">
            <Button onClick={decrement}>-</Button>
          </Tooltip>

          <span className={styles.count}>{count}</span>

          <Tooltip title="Increment" placement="top">
            <Button onClick={increment}>+</Button>
          </Tooltip>
        </div>

        <div className={styles.info}>{t('counter.info')}</div>
      </div>
    </>
  );
};

const mapStateToProps = createStructuredSelector<AppState, StateToProps>({
  count: makeSelectCount(),
});

const mapDispatchToProps: DispatchToProps = {
  increment: counterIncrement,
  decrement: counterDecrement,
};

export default connect<StateToProps, DispatchToProps, {}, AppState>(
  mapStateToProps,
  mapDispatchToProps,
)(CounterPage);
