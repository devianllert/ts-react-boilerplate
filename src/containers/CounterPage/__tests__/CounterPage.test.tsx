import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';

import { renderWithRouterAndRedux } from '../../../utils/testUtils';

import CounterPage from '../CounterPage';

import '@testing-library/jest-dom/extend-expect';

describe('<Counter />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const { container } = renderWithRouterAndRedux(<CounterPage />);

    expect(container).toMatchSnapshot();
  });

  it('should render with initial state', (): void => {
    const initialState = {
      counter: {
        count: 1000,
      },
    };

    const { getByText } = renderWithRouterAndRedux(<CounterPage />, { initialState });

    expect(getByText(/1000/)).toBeDefined();
  });

  it('should increment count', (): void => {
    const { getByText } = renderWithRouterAndRedux(<CounterPage />);

    expect(getByText(/0/)).toBeDefined();

    fireEvent.click(getByText('+'));

    expect(getByText(/1/)).toBeDefined();
  });

  it('should decrement count', (): void => {
    const { getByText } = renderWithRouterAndRedux(<CounterPage />);

    expect(getByText(/0/)).toBeDefined();

    fireEvent.click(getByText('-'));

    expect(getByText(/-1/)).toBeDefined();
  });
});
