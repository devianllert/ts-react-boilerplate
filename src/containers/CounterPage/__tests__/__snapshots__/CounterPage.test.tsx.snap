// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> should render and match the snapshot 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="counter"
    >
      <button
        class="button primary medium"
        title="Decrement"
        type="button"
      >
        -
      </button>
      <span
        class="count"
      >
        0
      </span>
      <button
        class="button primary medium"
        title="Increment"
        type="button"
      >
        +
      </button>
    </div>
    <div
      class="info"
    >
      counter.info
    </div>
  </div>
</div>
`;
